### PR TITLE
Sorting and Filtering

### LINKED ISSUE #19: Filter by Value
**Is your feature request related to a problem? Please describe.**
Allows the user to filter Pokemon by their value.

### PR DESCRIPTION
**Pull request by JayDampitan**
## Changes
1. Enable users to filter by type 
2. Enable users to sort by alphabetically, value and Id. 

## Purpose
Enhance user experience by allowing users to easily locate Pokemons.

## Approach
Used Enumerators and multiple functions to handle the data. 

Closes #19 , #20 , #22

### CODE REVIEW

File: pokeapp/src/components/FilterBar.js [hunk 1]
```diff
@@ -72,54 +93,136 @@ const [type, setType] = useState(false);
...
   };
 
+  // If there are two types only the pokemons with both clicked types will be returned
+  const filterByType = (pokemonList, type) => {
+    const sortByType = pokemonList.filter((pokemons) => {
+      for (const pokemon of pokemons.types) {
+        if (pokemon.type.name === type) {
+          return pokemon;
+        }
+      }
```
Comment (kmausolf) [inline pokeapp/src/components/FilterBar.js:103]: Good way of handling this, nice job.

File: pokeapp/src/components/FilterBar.js [hunk 2]
```diff
@@ -1,22 +1,45 @@
...
+  GHOST: "ghost",
+  GROUND: "ground",
+  ICE: "ice",
+  NORMAL: "normal",
+  POISON: "poison",
+  PSYCHIC: "psychic",
+  ROCK: "rock",
+  STEEL: "steel",
+  WATER: "water",
+};
```
Comment (kmausolf) [inline pokeapp/src/components/FilterBar.js:32]: We should extract these out to an `enums.js` file.

File: pokeapp/src/components/FilterBar.js [hunk 3]
```diff
@@ -72,54 +93,136 @@ const [type, setType] = useState(false);
...
+        {checkboxEl(ElementTypes.DARK, `Dark`)}
+        {checkboxEl(ElementTypes.DRAGON, `Dragon`)}
+        {checkboxEl(ElementTypes.ELECTRIC, `Electric`)}
+        {checkboxEl(ElementTypes.FAIRY, `Fairy`)}
+        {checkboxEl(ElementTypes.POISON, `Poison`)}
+        {checkboxEl(ElementTypes.FIGHTING, `Fighting`)}
+        {checkboxEl(ElementTypes.STEEL, `Steel`)}
+        {checkboxEl(ElementTypes.PSYCHIC, `Psychic`)}
+        {checkboxEl(ElementTypes.NORMAL, `Normal`)}
+        {checkboxEl(ElementTypes.ROCK, `Rock`)}
```
Comment (kmausolf) [inline pokeapp/src/components/FilterBar.js:225]: You can use [Object.entries()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries) to map over the enum, [then programmatically capitalize each first letter](https://stackoverflow.com/questions/1026069/how-do-i-make-the-first-letter-of-a-string-uppercase-in-javascript), or even better, have the enum contain the uppercased values and when you do your `pokemon.type.name` comparison, just use .toLowerCase().

File: pokeapp/src/utilities/ApiCall.js [hunk 4]
```diff
@@ -4,6 +4,7 @@ const P = new Pokedex.Pokedex();
 const interval = {
   offset: 0,
   limit: 1119,
+  cache: true,
```
Comment (kmausolf) [inline pokeapp/src/utilities/ApiCall.js:7]: Be sure to add in their service worker to allow for caching!

File: pokeapp/src/components/FilterBar.js (hunk 3 above)
Comment (krferrell) [reply pokeapp/src/components/FilterBar.js:225]: We will get on this on a new refactoring issue. Thanks for the idea!

File: pokeapp/src/components/FilterBar.js (hunk 2 above)
Comment (kmausolf) [reply pokeapp/src/components/FilterBar.js:32]: Wow, that code snippet got borked.  I mean to say that this file is big, and we should extract out some of the less directly related things.

**Review by Gjhnsn — APPROVED** [review]
looks good

File: pokeapp/src/Brain.js [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import React, { useState, useEffect } from "react";
-import { getAllPokemon, getPokemonWeakness } from "./utilities/ApiCall";
+import { getAllPokemon, getPokemonWeakness } from "./utilities/apiCall";
```
Comment (BlackCubes) [inline pokeapp/src/Brain.js:2]: The little things.........

File: pokeapp/src/components/FilterBar.js [hunk 6]
```diff
@@ -72,54 +63,136 @@ const [type, setType] = useState(false);
...
+        filteredPokemon = sortIdLow(filteredPokemon);
+        break;
+      case SortTypes.VALUE_HIGH_TO_LOW:
+        filteredPokemon = sortValueHigh(filteredPokemon);
+        break;
+      case SortTypes.VALUE_LOW_TO_HIGH:
+        filteredPokemon = sortValueLow(filteredPokemon);
+        break;
+      default:
+        console.error(`sort type ${sortType} is not a valid sort type`);
```
Comment (BlackCubes) [inline pokeapp/src/components/FilterBar.js:108]: You guys are going HAM!!!

File: pokeapp/src/components/FilterBar.js [hunk 7]
```diff
@@ -72,54 +63,136 @@ const [type, setType] = useState(false);
...
     filterPokemon();
-  },[sortType, type])
+  }, [sortType, type, selectedTypes]);
+
+
+  const addRemoveTypeFilter = (elementType) => {
+    if (!selectedTypes.includes(elementType)) {
+      // If its not in the array add it to the array
+      setSelectedTypes((prevSelectedTypes) => [
+        ...prevSelectedTypes,
```
Comment (BlackCubes) [inline pokeapp/src/components/FilterBar.js:122]: Good functional programming here

**Review by BlackCubes — APPROVED** [review]
Everything looks good, and you guys are working hard! LET'S GO 🐱‍🏍

**Pull request merged**